base/abstractarray.jl

```diff
@@ -1962,6 +1981,258 @@ function typed_hvcat(::Type{T}, rows::Tuple{Vararg{Int}}, as...) where T
+for block matrix syntax. The first argument specifies the number of arguments in `values`
+along the first axis for each dimension. `row_first` indicates whether `values` are ordered
+first by rows (true) or first by columns (false).
+
+# Examples
+```jldoctest
+julia> a, b, c, d, e, f = 1, 2, 3, 4, 5, 6
+(1, 2, 3, 4, 5, 6)
+
+julia> [a b c;;; d e f]
+1×3×2 Array{Int64, 3}:
+[:, :, 1] =
+ 1  2  3
+
+[:, :, 2] =
+ 4  5  6
+
+julia> hvncat((2,1,3), false, a,b,c,d,e,f)
+2×1×3 Array{Int64, 3}:
+[:, :, 1] =
+ 1
+ 2
+
+[:, :, 2] =
+ 3
+ 4
+
+[:, :, 3] =
+ 5
+ 6
+
+julia> [a b;;; c d;;; e f]
+1×2×3 Array{Int64, 3}:
+[:, :, 1] =
+ 1  2
+
+[:, :, 2] =
+ 3  4
+
+[:, :, 3] =
+ 5  6
+```
+"""
+hvncat(dims::Tuple{Vararg{Int}}, row_first::Bool, xs::AbstractArray...) = typed_hvncat(promote_eltype(xs...), dims, row_first, xs...)
+hvncat(dims::Tuple{Vararg{Int}}, row_first::Bool, xs::AbstractArray{T}...) where T = typed_hvncat(T, dims, row_first, xs...)
+
+function typed_hvncat(::Type{T}, dims::Tuple{Vararg{Int, N}}, row_first::Bool, as::AbstractArray...) where T where N
+    N == 1 && return typed_vcat(T, xs...)
+    row_first && N == 2 && return typed_hvcat(T, ntuple(x->dims[2], length(xs) ÷ dims[2]), as...)
+
+    d1 = row_first ? 2 : 1
+    d2 = row_first ? 1 : 2
+
+    # discover dimensions
+    nd = max(N, ndims(as[1]))
+    outdims = zeros(Int, nd)
+
+    # discover number of rows/columns
+    @inbounds for i ∈ 1:dims[d1]
+        outdims[d1] += size(as[i], d1)
+    end
+
+    currentdims = zeros(Int, nd)
+    blockcount = 0
+    @inbounds for i ∈ eachindex(as)
+        currentdims[d1] += size(as[i], d1)
+        if currentdims[d1] == outdims[d1]
+            currentdims[d1] = 0
+            for d ∈ (d2, 3:nd...)
+                currentdims[d] += size(as[i], d)
+                if outdims[d] == 0 # unfixed dimension
+                    blockcount += 1
+                    if blockcount == (d > length(dims) ? 1 : dims[d]) # last expected member of dimension
+                        outdims[d] = currentdims[d]
+                        currentdims[d] = 0
+                        blockcount = 0
+                    else
+                        break
+                    end
+                else # fixed dimension
+                    if currentdims[d] == outdims[d] # end of dimension
+                        currentdims[d] = 0
+                    elseif currentdims[d] < outdims[d] # dimension in progress
+                        break
+                    else # exceeded dimension
+                        ArgumentError("Incompatible dimension in argument $(i) along axis $(d)") |> throw
+                    end
+                end
+            end
+        end
+    end
+
+    len = sum(length, as)
+    outlen = prod(outdims) # extracted to avoid hitting allocation bug
+    len == outlen || ArgumentError("mismatch between length of inferred output dimensions and input arrays; expected $(outlen), got $(len)") |> throw
```

Comment:
In one of the tests I did, this error message was odd:

```Julia
julia> [fill(1,1);fill(1,1);;fill(1, 3);;rand(3,3)]
ERROR: ArgumentError: mismatch between length of inferred output dimensions and input arrays; expected 0, got 14
```

Reply:
Good catch! Would this be better? "argument 3 has too many elements along axis 1"